pollux/lib/core/src/main/scala/io/iohk/atala/pollux/core/service/PresentationServiceImpl.scala

```diff
@@ -401,7 +440,96 @@ private class PresentationServiceImpl(
+                  anoncredSchema.attrNames,
+                  anoncredSchema.issuerId
+                )
+            } yield (schemaId, anoncredLibSchema)
+          })
+          .map(_.toMap)
+      credentialDefinitionMap <-
+        ZIO
+          .collectAll(credentialDefinitionIds.map { credentialDefinitionId =>
+            for {
+              credentialDefinition <- credentialDefinitionService
+                .getByGUID(credentialDefinitionId)
+                .mapError(e => UnexpectedError(e.toString))
+            } yield (credentialDefinition.longId, CredentialDefinition(credentialDefinition.definition.toString))
+          })
+          .map(_.toMap)
+
+      verifiableCredentials <-
+        ZIO.collectAll(
+          issuedCredentials
+            .flatMap(_.attachments)
+            .filter(attachment => attachment.format.contains(IssueCredentialIssuedFormat.Anoncred.name))
+            .map(_.data)
+            .map {
+              case Base64(data) => Right(new String(JBase64.getUrlDecoder.decode(data)))
+              case _            => Left(InvalidAnoncredPresentationRequest("Expecting Base64-encoded data"))
+            }
+            .map(ZIO.fromEither(_))
+        )
+      presentationRequestAttachment <- ZIO.fromEither(
+        requestPresentation.attachments.headOption.toRight(InvalidAnoncredPresentationRequest("Missing Presentation"))
+      )
+      presentationRequestData <-
+        presentationRequestAttachment.data match
+          case Base64(data) => ZIO.succeed(new String(JBase64.getUrlDecoder.decode(data)))
+          case _            => ZIO.fail(InvalidAnoncredPresentationRequest("Expecting Base64-encoded data"))
+      deserializedPresentationRequestData <-
+        AnoncredPresentationRequestV1.schemaSerDes
+          .deserialize(presentationRequestData)
+          .mapError(error => InvalidAnoncredPresentationRequest(error.error))
+      linkSecret <-
+        linkSecretService
+          .fetchOrCreate()
+          .map(_.secret)
+          .mapError(t => AnoncredPresentationCreationError(t.cause))
+      presentation <-
+        ZIO
+          .fromEither(
+            Try(
+              AnoncredLib.createPresentation(
+                PresentationRequest(presentationRequestData),
+                verifiableCredentials.map(verifiableCredential =>
+                  CredentialAndRequestedAttributesPredicates(
+                    Credential(verifiableCredential),
+                    deserializedPresentationRequestData.requested_attributes.keys.toSeq, // TO FIX
+                    deserializedPresentationRequestData.requested_predicates.keys.toSeq // TO FIX
+                  )
+                ),
+                Map.empty, // TO FIX
+                linkSecret,
+                schemaMap,
+                credentialDefinitionMap
+              )
+            ).toEither
+          )
+          .mapError((t: Throwable) => AnoncredPresentationCreationError(t))
+    } yield presentation
   }
```

Comment:
the method is tooooo long :)
try to refactor it to a sequence of smaller functions next time.

Reply:
noted